### PR TITLE
RFC: Refactor/paginator

### PR DESCRIPTION
**Pull request by burzum**
There is no reason why the pagination component has  to be coupled to the component / controller layer. If you want to use pagination inside a shell or a service layer you'll have two choices:
- Ignore MVC and import the paginator component in a place where it doesn't belong to technically
- Re-invent the wheel and implement your own pagination

This change moved the whole pagination logic into a new class.

There is one thing to discuss: I couldn't find a better way to implement the pagination based on the request than extending the new class by another class and injecting the request there. Ideas how to do this better are welcome! The paginator component test passes for me with my current changes.

### CODE REVIEW

File: src/ORM/Paginator.php [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         2.0.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM;
+
+use Cake\Controller\Component;
```
Comment (jippi) [inline src/ORM/Paginator.php]: this will create inter-package dependencies with the ORM sub-tree split

Comment (burzum) [inline src/ORM/Paginator.php]: True, forgot to remove that, did it now.

File: src/Controller/Component/PaginatorComponent.php [hunk 2]
```diff
@@ -65,305 +44,43 @@ public function implementedEvents()
...
-        }
-        $request['paging'] = [$alias => $paging] + (array)$request['paging'];
-
-        if ($requestedPage > $page) {
-            throw new NotFoundException();
-        }
-
-        return $results;
+        $this->_paginator = new Paginator($config);
+        $this->_paginator->request = $registry->getController()->request;
```
Comment (markstory) [inline src/Controller/Component/PaginatorComponent.php]: Looks the request should be a constructor argument instead.

**Comment by markstory** [comment]
Seems reasonable to me. Having two classes that start with Paginator in Controller/Component is a bit confusing though.

File: src/Controller/Component/PaginatorComponent.php (hunk 2 above)
Comment (burzum) [inline src/Controller/Component/PaginatorComponent.php]: The idea was to remove depedencies from it. If you don't like to access the property directly, maybe implement a setter?

**Comment by burzum** [comment]
@markstory any better ideas?

**Comment by markstory** [comment]
Would it be possible to put the code in Paginator into the PaginatorComponent? Having two things depend on the request seems somewhat redundant. The OrmPaginator separates the bulk of pagination out of the component/request bound context which is nice.

**Comment by burzum** [comment]
@markstory please check my last commit. This should resolve the need to ever extend the paginator. :smile:  At least I hope so. Instead of messing with the request object I've implemented what I wanted to do from the very beginning. A setter for the params that get merged in mergeOptions().

This allows the paginator component to be BC and other pagination things to work very similar: Just pass your array of options that are coming from somewhere to setParams() and you're basically done with pagination.

File: src/Controller/Component/PaginatorComponent.php [hunk 3]
```diff
@@ -65,305 +44,60 @@ public function implementedEvents()
...
-    public function checkLimit(array $options)
+    public function __call($method, $args)
     {
-        $options['limit'] = (int)$options['limit'];
-        if (empty($options['limit']) || $options['limit'] < 1) {
-            $options['limit'] = 1;
+        if (method_exists($this->_paginator, $method)) {
+            return call_user_func_array([$this->_paginator, $method], $args);
         }
-        $options['limit'] = min($options['limit'], $options['maxLimit']);
```
Comment (markstory) [inline src/Controller/Component/PaginatorComponent.php]: If the method doesn't exist you should raise an exception. Silent failures are not very nice.

File: tests/TestCase/Controller/Component/PaginatorComponentTest.php [hunk 4]
```diff
@@ -379,6 +382,7 @@ public function testMergeOptionsExtraWhitelist()
             'conditions' => ['bad.stuff'],
             'contain' => ['bad']
         ];
+		$this->Paginator->setParams($this->request->query);
```
Comment (markstory) [inline tests/TestCase/Controller/Component/PaginatorComponentTest.php]: Indentation seems funny.

File: src/ORM/Paginator.php [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         2.0.0
```
Comment (ADmad) [inline src/ORM/Paginator.php]: 3.1.0

File: src/Controller/Component/PaginatorComponent.php [hunk 6]
```diff
@@ -65,305 +44,60 @@ public function implementedEvents()
...
 
-    /**
-     * Extracts the finder name and options out of the provided pagination options
-     *
-     * @param array $options the pagination options
-     * @return array An array containing in the first position the finder name and
-     * in the second the options to be passed to it
-     */
-    protected function _extractFinder($options)
+    public function paginate($object, array $settings = [])
```
Comment (ADmad) [inline src/Controller/Component/PaginatorComponent.php]: API docblock missing.

File: src/Controller/Component/PaginatorComponent.php [hunk 7]
```diff
@@ -65,305 +44,60 @@ public function implementedEvents()
...
-     *   return $q->where(['name' => 'CakePHP'])
-     * });
-     * $results = $paginator->paginate($query);
-     * ```
-     *
-     * @param \Cake\Datasource\RepositoryInterface|\Cake\ORM\Query $object The table or query to paginate.
-     * @param array $settings The settings/configuration used for pagination.
-     * @return array Query results
-     * @throws \Cake\Network\Exception\NotFoundException
+     * @param ComponentRegistry $registry A ComponentRegistry this component can use to lazy load its components
```
Comment (ADmad) [inline src/Controller/Component/PaginatorComponent.php]: Please use full namespace.

File: src/ORM/Paginator.php [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+     * If you need to sort on associated columns or synthetic properties you will need to use a whitelist.
+     *
+     * Any columns listed in the sort whitelist will be implicitly trusted. You can use this to sort
+     * on synthetic columns, or columns added in custom find operations that may not exist in the schema.
+     *
+     * @param Table $object The model being paginated.
+     * @param array $options The pagination options being used for this request.
+     * @return array An array of options with sort + direction removed and replaced with order if possible.
+     */
+    public function validateSort(Table $object, array $options)
```
Comment (ADmad) [inline src/ORM/Paginator.php]: Maybe the typehint `Table` should be changed to `RepositoryInterface` to potentially allow more object types in future.

File: src/ORM/Paginator.php [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+        'maxLimit' => 100,
+        'whitelist' => ['limit', 'sort', 'page', 'direction']
+    ];
+
+    /**
+     * Pagination parameters.
+     *
+     * @var array
+     */
+    public $_pagingParams = [];
```
Comment (ADmad) [inline src/ORM/Paginator.php]: Should be `protected`.

File: src/ORM/Paginator.php [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+
+    /**
+     * Params that get merged in the next paginate() call.
+     *
+     * @var array
+     */
+    public $_params = [];
```
Comment (ADmad) [inline src/ORM/Paginator.php]: Should be `protected`.

File: tests/TestCase/Controller/Component/PaginatorComponentTest.php (hunk 4 above)
Comment (burzum) [inline tests/TestCase/Controller/Component/PaginatorComponentTest.php]: Yea, fuck spaces and PSR... :disappointed:

File: src/ORM/Paginator.php (hunk 8 above)
Comment (burzum) [inline src/ORM/Paginator.php]: I've tried that and got this:

```
1) Cake\Test\TestCase\Controller\Component\PaginatorComponentTest::testPaginateQuery
Argument 1 passed to Cake\ORM\Paginator::paginate() must implement interface Cake\Datasource\RepositoryInterface, instance of Mock_Query_143691e0 given,
```

What's the recommended way to resolve this?

File: src/Controller/Component/PaginatorComponent.php (hunk 3 above)
Comment (burzum) [inline src/Controller/Component/PaginatorComponent.php]: What method? Did you comment on the right line? If yes I don't get the problem here.

Comment (markstory) [inline src/Controller/Component/PaginatorComponent.php]: I was trying to comment on the __call method that silently fails when methods are missing.

File: src/ORM/Paginator.php (hunk 8 above)
Comment (ADmad) [inline src/ORM/Paginator.php]: Seems the typehint should be removed. `$object` can be either instance of  `Query` or instance of class implementing `RepositoryInterface` as stated in docblock of `paginate()`.

File: src/Controller/Component/PaginatorComponent.php (hunk 3 above)
Comment (burzum) [inline src/Controller/Component/PaginatorComponent.php]: Ahh OK, I agree.

**Comment by burzum** [comment]
What's left before this can be merged?

**Comment by markstory** [comment]
Lots of [build failures](https://travis-ci.org/cakephp/cakephp/builds/60295161].

**Comment by burzum** [comment]
I've already tried to fix this and I think this is a bug in the ObjectRegistry. 

https://travis-ci.org/cakephp/cakephp/jobs/60295163

Going to look at it again the next time I have time available.

**Comment by markstory** [comment]
Closing as this cannot be merged and still needs a bit of work. If you make progress on this in the future please reopen :smile:

**Pull request closed without merging**